File: October2022/Java/CountAndSay.java
// @saorav21994

class Solution {
    
    public String calc(String s) {
        char ch = s.charAt(0);
        int count = 1;
        int l = s.length();
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < l; i++) {
            if(s.charAt(i) != ch) {
                sb.append(count).append(ch);
                ch = s.charAt(i);
                count = 1;
            }
            else {
                count += 1;
            }
        }
        sb.append(count).append(ch);
        return sb.toString();
    }
    
    public String countAndSay(int n) {
        if (n == 1) return "1";
        String res = "1";
        for (int i = 2; i <= n; i++) {
            res = calc(res);
        }
        return res;
    }
}

File: October2022/Python/CountAndSay.py
# https://leetcode.com/problems/count-and-say/
# Author : Jiganesh

class Solution:
    
    # TC : O(N^2) # 30N 
    # SC :O(N)
    def countAndSay(self, n: int) -> str:

        def helper(string):
            ch = string[0]
            count = 0

            result = ""
            for i in string:
                if ch == i:
                    count+=1
                else:
                    result+= str(count) + ch
                    ch = i
                    count=1
            result+= str(count) + ch
            return result

        string = "1"

        if n == 1:
            return "1"
            
        while n>1:
            print(string)
            string = helper(string)
            n-=1
        return string
